SYCL/Matrix/joint_matrix_nvptx_double.cpp

```diff
@@ -0,0 +1,152 @@
+constexpr int SUB_TILES_N =
+    55; // number of submatrices per col of accumulator matrices.
+constexpr int SUB_TILES_K =
+    257; // number of submatrices per col of "A"/per row of "B", matrices.
+
+constexpr int BIG_M =
+    SUB_TILES_M *
+    M; // total number of M dimension matrix elements for the "Big matrix".
+constexpr int BIG_N =
+    SUB_TILES_N *
+    N; // total number of N dimension matrix elements for the "Big matrix".
+constexpr int BIG_K =
+    SUB_TILES_K *
+    K; // total number of K dimension matrix elements for the "Big matrix".
+
+// The stride should equal the number of elements between consecutive leading
+// dimensions of the "Big matrix". e.g. number of elements per row if matrix is
+// indexed row major. The stride tells the implementation how many elements to
+// skip in memory matrix row/column multiplications.
+constexpr int STRIDE_A = BIG_K; // row major. If col major should equal BIG_M.
+constexpr int STRIDE_B = BIG_N; // row_major. If col major should equal BIG_K.
+constexpr int STRIDE_C = BIG_N; // row major. If col major should equal BIG_M.
+
+double A[BIG_M * BIG_K];
+double B[BIG_K * BIG_N];
+double C[BIG_M * BIG_N];
+double D[BIG_M * BIG_N];
+
+// returns correct (m,n) element of matrix D = A*B + C (assuming all matrices
+// are indexed row_major).
+double matrix_ref_mn(const int &m, const int &n) {
+  double res = C[m * BIG_N + n];
+
+  for (int k = 0; k < BIG_K; k++)
+    res += A[m * BIG_K + k] * B[k * BIG_N + n];
+  return res;
+}
+
+int main() {
+  for (int i = 0; i < BIG_M * BIG_N; i++) {
+    C[i] = i;
+    D[i] = 0;
+  }
+
+  for (int i = 0; i < BIG_M * BIG_K; i++) {
+    A[i] = i;
+  }
+
+  for (int i = 0; i < BIG_K * BIG_N; i++) {
+    B[i] = i;
+  }
+
+  buffer<double, 1> bufA(A, range<1>(BIG_M * BIG_K));
+  buffer<double, 1> bufB(B, range<1>(BIG_K * BIG_N));
+  buffer<double, 1> bufC(C, range<1>(BIG_M * BIG_N));
+  buffer<double, 1> bufD(D, range<1>(BIG_M * BIG_N));
+
+  queue q;
+  q.submit([&](handler &cgh) {
+    auto accC = bufC.get_access<access::mode::read_write>(cgh);
+    auto accA = bufA.get_access<access::mode::read_write>(cgh);
+    auto accB = bufB.get_access<access::mode::read_write>(cgh);
+    auto accD = bufD.get_access<access::mode::read_write>(cgh);
+
+    range<2> LocalRange = {1, N_THREADS_PER_MATRIX_OP};
+    range<2> GlobalRange = {SUB_TILES_M, SUB_TILES_N * N_THREADS_PER_MATRIX_OP};
+
+    cgh.parallel_for<class imatrix>(
+        nd_range<2>(GlobalRange, LocalRange), [=
+    ](nd_item<2> item) [[sycl::reqd_work_group_size(1, 1, 32)]] {
+          sycl::sub_group sg = item.get_sub_group();
+
+          const auto m =
+              item.get_group()
+                  .get_id()[0]; // row id of current submatrix of BIG C matrix
+          const auto n =
+              item.get_group().get_id()[1]; // column id of current submatrix of
+                                            // BIG C matrix
+
+          joint_matrix<double, matrix_use::accumulator, M, N,
```

Comment:
I think a comment is necessary here to highlight the difference with the current matrix interface WRT the use of the "use" argument.

Reply:
OK

Reply:
Made the change